src/lib/sql/sql_translator.cpp

```diff
@@ -1799,16 +1803,33 @@ std::shared_ptr<AbstractExpression> SQLTranslator::_translate_hsql_expr(
 
     case hsql::kExprCast: {
       const auto source_data_type = left->data_type();
-      if (expr.columnType.data_type == hsql::DataType::DATE) {
-        AssertInput(source_data_type == DataType::String, "Cannot cast " + left->as_column_name() + " as Date");
-        // We do not know if an expression to be casted other than a ValueExpression actually contains dates, and we
-        // cannot check this later due to the lack of a Date data type
-        AssertInput(left->type == ExpressionType::Value, "Only ValueExpressions can be casted as Date");
-        const auto date_string = boost::get<pmr_string>(static_cast<ValueExpression&>(*left).value);
-        const auto date = string_to_date(std::string{date_string});
-        if (date) return std::const_pointer_cast<AbstractExpression>(left);
-        FailInput("'" + std::string{date_string} + "' is not a valid date");
+      const auto target_hsql_data_type = expr.columnType.data_type;
+
+      if (target_hsql_data_type == hsql::DataType::DATE || target_hsql_data_type == hsql::DataType::DATETIME) {
+        AssertInput(source_data_type == DataType::String,
+                    "Cannot cast " + left->as_column_name() + " as " +
+                        std::string{magic_enum::enum_name(target_hsql_data_type)});
+        // We do not know if an expression to be casted other than a ValueExpression actually contains date time
+        // values, and we cannot check this later due to the lack of proper data types.
+        AssertInput(left->type == ExpressionType::Value, "Only ValueExpressions can be casted as " +
+                                                             std::string{magic_enum::enum_name(target_hsql_data_type)});
+        const auto input_string = std::string{boost::get<pmr_string>(static_cast<ValueExpression&>(*left).value)};
+
+        if (target_hsql_data_type == hsql::DataType::DATE) {
+          const auto date = string_to_date(input_string);
+          AssertInput(date, "'" + input_string + "' is not a valid date");
+          return std::const_pointer_cast<AbstractExpression>(left);
```

Comment:
Besides that I am kind of confused that we don't to any casting here. We throw away `date` and that's it?

Reply:
Absolutely correct, the `cast` is not necessary.

Well, we don't have a `Date` data type, so we just check if the given string is a correct date and return the string's value expression.

Reply:
Added a comment here

Reply:
> `left` is a n abstract expression, right? Why the cast?
Argh, the cast was there for a reason...
```
/var/lib/jenkins/workspace/hyrise_hyrise_PR-2471/src/lib/sql/sql_translator.cpp:1822:18: error: constness of 'left' prevents automatic move [performance-no-automatic-move,-warnings-as-errors]

          return left;
```

So I *could* remove the `const` qualifier from the variable declaration, do the cast, or return `left->deep_copy()`. Opinions?

Reply:
Haven't found the code I was commenting on. Can we resolve? I am confused. :D

Reply:
Still there somehow, the line is:
https://github.com/hyrise/hyrise/blob/dey4ss/datetime/src/lib/sql/sql_translator.cpp#L1850

I removed the `const` qualifier or the `left` variable, which is one possible solution. The `CAST` expression also returns the left input in L1869. I dod not remove the `const_pointer_cast` there since I was not sure what was your opinion on how we should return the left input.

Reply:
Feel free to remove the other const cast as well.